### ISSUE TITLE
'test' profile is used

### ISSUE THREAD

**Issue by eproxus** [crownedgrouse/geas_rebar3#5]
It seems the plugin uses the `test` profile (analyzing code in `_build/test`). I think it should use the `default` profile instead, because test beams are not always compiled and up to date.

**Comment by crownedgrouse** [comment]
In fact, everything under _build is analyzed I guess. I'm more erlang.mk guy, so I need some introspection on this before fixing. I will ask your help if needed.

**Comment by eproxus** [comment]
True, I can see my modules listed several times in the right hand side column.

I think it should be fine just descending into `_build/default/lib` and do the analysing there. This would also avoid analysing `_build/default/plugins` including Geas itself 😊

**Comment by eproxus** [comment]
You might also be able to get the current profile from Rebar and use that to discover the path. Rebar 3 can be run as `rebar as test geas` which means the `test` profile is used (which Geas currently ignores as you explained).

**Comment by eproxus** [comment]
After some more research: `Profiles = rebar_state:current_profiles(State)` would give you a list of the currently active profile. By default it is `[default]` and if Geas is invoked as `rebar3 as test geas` it is `[default, test]`. Not sure what is bes there. Using `lists:last/1` or removing `default` if the list is longer than 1 might be an option.

**Comment by eproxus** [comment]
Wanted to do a pull request, but the current Geas api only allows for one directory to be scanned 😁

This is fine in the default case, but would not allow for easily implementing `rebar3 as profile1,profile2,profile3 geas`. Suggestions?

**Comment by crownedgrouse** [comment]
for now I suppose _build/default  might sufficient. I need to think more on this from your comments in order to handle several profile. but the idea is interesting. feel free to propose a PR for default only

**Comment by eproxus** [comment]
@crownedgrouse Will do a PR for default for now 👍

**Comment by crownedgrouse** [comment]
See line 1107 in geas.erl

**Comment by eproxus** [comment]
Ouch, so Geas always adds `_build` to the dir? I assumed it would be possible to do `geas:compat(".../_build/default/lib")` but would it then analyse `".../_build/default/lib/_build"`?

**Comment by crownedgrouse** [comment]
_build is the root directory to search the modules below when using it for a global project. change to _build/default  show do the trick for now, while something more smart.

**Comment by crownedgrouse** [comment]
_build is not always added, only if deps not found if I remember. should work for erlang.mk and rebar

**Comment by eproxus** [comment]
@crownedgrouse Ah, you mean that the fix has to go to Geas and not geas_rebar3? Shouldn't Geas be unopiniated about where to look, and just accept the path argument as is?

**Comment by crownedgrouse** [comment]
Geas, as plugin  need to detect where it runs, and in what kind of project (erlang.mk or rebar), it do it by searching deps/ or _build/  directory. Geas as module can receive a directory path as argument. Both plugins call geas module without current pwd .

**Comment by crownedgrouse** [comment]
it is why I use a lot of env variable. We could think to set an env variable for profile, and append it to _build directory.

**Comment by eproxus** [comment]
But if I change the geas_rebar3 plugin to detect which folder to run in, Geas will break this by adding extra folders to the path? In my opinion, I think Geas should just check the path given to it, and it is up to the erlang.mk and Rebar 2/3 plugins to set the special path needed for that build environment.

**Comment by crownedgrouse** [comment]
I need to think about it, before breaking things.

**Comment by eproxus** [comment]
Of course 😊

**Comment by crownedgrouse** [comment]
what could be helpful , as I'm not a rebar guy, is to give me a project sample with profile I could use. Maybe yours ?

**Comment by eproxus** [comment]
This will give you an example project (given geas_rebar3 installed as a global plugin in `~/.config/rebar3/rebar.config`):

```sh
$ rebar3 new app name="foo"
===> Writing foo/src/foo_app.erl
===> Writing foo/src/foo_sup.erl
===> Writing foo/src/foo.app.src
===> Writing foo/rebar.config
===> Writing foo/.gitignore
===> Writing foo/LICENSE
===> Writing foo/README.md
$ cd foo
$ rebar3 do compile, eunit
===> Verifying dependencies...
===> Compiling foo
===> Verifying dependencies...
===> Compiling foo
===> Performing EUnit tests...

Finished in 0.055 seconds
0 tests
$ rebar3 geas
===> Verifying dependencies...
===> Compiling foo
   R15                   19.3       Geas database
---Min--------Arch-------Max----------------------------------------------------
                                    foo
                                    foo
                                    foo
--------------------------------------------------------------------------------
   R15                   19.3       Global project
```

As you can see, foo is listed 3 times, instead of ideally only 1 time.

**Comment by crownedgrouse** [comment]
Thanks. I understand better why I did this, because geas can work either on beam or source file depending env var. It might better to simply make the application list uniq (only one time 'foo').

**Comment by eproxus** [comment]
The problem is, depending on where you look in Rebar 3's build structure (e.g. in `_build/default` or in `_build/test`) you'll get different beam files, which might be relevant to the user. Perhaps they have a special production profile with special compiler macros that replace function calls and want to see if it is compatible with a certain Erlang release using Geas.

**Comment by crownedgrouse** [comment]
See new geas version 2.6.1 and rebar3 plugin 1.3.1 .  it is better ?  Next version should be able to deal with rebar3 configs. on the roadmap.

**Comment by eproxus** [comment]
I tried it again and it still seems to list the project twice:
```
$ rebar3 geas
===> Fetching geas_rebar3 v1.3.1
===> Downloaded package, caching at /Users/user/.cache/rebar3/hex/hexpm/packages/geas_rebar3-1.3.1.tar
===> Fetching geas v2.6.1
===> Downloaded package, caching at /Users/user/.cache/rebar3/hex/hexpm/packages/geas-2.6.1.tar
===> Fetching samovar v1.0.0
===> Downloaded package, caching at /Users/user/.cache/rebar3/hex/hexpm/packages/samovar-1.0.0.tar
===> Compiling samovar
===> Compiling geas
===> Compiling geas_rebar3
===> Verifying dependencies...
===> Compiling foo
   R15                   22.2       Geas database                       2.6.1
---Min--------Arch-------Max----------------------------------------------------
                                    foo                                 0.1.0
                                    geas_rebar3                         1.3.1
                                    foo                                 0.1.0
--------------------------------------------------------------------------------
   R15                   22.2       Global project                      0.1.0

C : 22.1.7
L : R15 R15B R15B01 R15B02 R15B03 R15B03-1 R16B R16B01 R16B02 R16B03 R16B03-1 17.0 17.1 17.3 17.4 17.5 18.0 18.1 18.2 18.3 19.0 19.1 19.2 19.3 20.0 20.1 20.2 20.3 21.0 21.1 21.2 22.0 22.1 22.2
```

**Comment by crownedgrouse** [comment]
Can you migrate your Geas_rebar3 plugin declaration to global rebar3 config instead of local rebar config?
If you have it twice and two profiles, normal for now as profiles will be handled in next version.

**Comment by crownedgrouse** [comment]
I added this note in project's readme : Important : build plugins must not be set in local rebar.config. Otherwise those plugins will be part of geas analyze.

**Comment by eproxus** [comment]
Another note: there's now a `{project_plugins, [geas_rebar3]}` configuration for Rebar that I think should be used for this purpose. This will install those apps and their dependencies to `_build/default/plugins`. Perhaps this folder could be ignored when running the Geas plug-in for Rebar?
